synapse/storage/databases/main/events_worker.py

```diff
@@ -736,35 +736,73 @@ async def get_stripped_room_state_from_event_context(
             for e in state_to_include.values()
         ]
 
-    def _do_fetch(self, conn: Connection) -> None:
+    async def _do_fetch(self) -> None:
+        """Services requests for events from the `_event_fetch_list` queue."""
+        try:
+            await self.db_pool.runWithConnection(self._do_fetch_txn)
+        except Exception as e:
+            failed_event_list = []
+            with self._event_fetch_lock:
+                if self._event_fetch_ongoing == 1 and self._event_fetch_list:
+                    # We are the last remaining fetcher and we have just failed.
+                    # Fail any outstanding fetches, since they won't get processed
+                    # otherwise.
+                    failed_event_list = self._event_fetch_list
+                    self._event_fetch_list = []
+
+                self._event_fetch_ongoing -= 1
+                event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
+
+            for _, deferred in failed_event_list:
+                if not deferred.called:
+                    with PreserveLoggingContext():
+                        deferred.errback(e)
+
+            raise
+        else:
+            should_restart = False
+            with self._event_fetch_lock:
+                if self._event_fetch_ongoing == 1 and self._event_fetch_list:
+                    # An event fetch has been queued, but we're the last remaining
+                    # fetcher and have already decided to terminate.
+                    # Start a new fetcher.
+                    should_restart = True
+
+                    # `_event_fetch_ongoing` ought to be decremented for ourself and
+                    # incremented for the new fetcher. These cancel out, so we leave it
+                    # alone.
+                else:
+                    self._event_fetch_ongoing -= 1
+                    event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
```

Comment:
Can this happen in a single place please? Probably in a finally block? I do wonder if maybe we should move this into the same function that increments the `self._event_fetch_ongoing` in the first place?

Reply:
I had a go at refactoring the code. Let me know if it looks alright!